feat: add new_transaction support

daniel-sanche opened the pull request:
This PR modifies the Transaction class to add a new `begin_later` argument. 

When True, the transaction will not initialize itself at the start of the context manager `__enter__` block. Instead, it will only begin at the first rpc. If the first rpc is a read call, this allows us to create the transaction at the same time as the call, saving a network call.

This new behaviour is enabled by default, and is aligned with changes in other languages

---
## For reviewers

In Python, a transaction is created in a context manager, and typically interacted with through the client (which passes through to the Transaction object when needed):

```
with client.transaction(begin_later=True):
  client.query(query)
  client.get(entity)
  client.put(entity)
  client.delete(entity)
```

if begin_later is False, it will call `begin` automatically when entering that context manager block (using \_\_enter\_\_ method), retrieving a transaction_id from the server. Otherwise, begin is deferred until a get/query/commit call forces the transaction to talk to the backend

### States
The state machine looks like this. Green lines denote transitions that are only valid when `begin_later=True` orange lines denote transitions that are valid when `begin_later=False`

![Untitled Diagram drawio (2)](https://github.com/googleapis/python-datastore/assets/3914119/720a2ba4-5242-478f-b966-66e1b433bcfd)

- Starts out in INITIAL state
  - if begin_later is true, it allows you to immediately start collecting  put and get mutations. Otherwise any rpcs in the INITIAL state will throw an exception
- When `begin` is called and the transaction receives and ID from the server, it moves into IN_PROGRESS state
   -  if begin_later is false, it will immediately call begin when __enter__ is called (called by the `with` statement)
   - if begin_later is true, it will call begin when it encounters the first `get` or `query` call
   - user can also always call `begin` manually
- a rollback from either state (or a commit with no content from INITIAL) will result in the terminal ABORTED state
- a commit from IN_PROGRESS results in the terminal FINISHED state
- any invalid attempted transitions will raise an exception

daniel-sanche committed:
added begin_later flag

daniel-sanche committed:
refactoring

daniel-sanche committed:
got tests passing

daniel-sanche committed:
improved docstring

daniel-sanche committed:
added helper tests

daniel-sanche committed:
fixed type in docstring

daniel-sanche committed:
added client test

daniel-sanche committed:
handle context manager close

daniel-sanche committed:
removed extra indents

daniel-sanche committed:
added transaction tests

daniel-sanche committed:
added system tests

daniel-sanche committed:
fixed lint

daniel-sanche committed:
refactor using wrappers

daniel-sanche committed:
added functools.wraps

daniel-sanche committed:
fixed mypy

daniel-sanche committed:
Merge branch 'main' into new_transaction

daniel-sanche committed:
fixed mypy

daniel-sanche commented on google/cloud/datastore/transaction.py:
```diff
@@ -149,15 +151,23 @@ class Transaction(Batch):
     _status = None
 
-    def __init__(self, client, read_only=False, read_time=None):
+    def __init__(self, client, read_only=False, read_time=None, begin_later=True):
```
Just wanted to confirm that we did want this enabled by default?

It shouldn't be a breaking change in most cases, but I can think of a few edge cases where behaviors might end up different after this change

daniel-sanche replied:
After discussing with Baha, defaulting to False as the other languages are doing

daniel-sanche committed:
added test

daniel-sanche commented on google/cloud/datastore/transaction.py:
```diff
@@ -240,6 +250,57 @@ def begin(self, retry=None, timeout=None):
+                self.begin()
+            return fn(self, *args, **kwargs)
+
+        return wrapped
```
Maybe we should make the public `id` getter a `_begin_if_not_began` function? Otherwise, existing code will start returning None where it used to have an ID. Thoughts?

daniel-sanche replied:
N/A, begin_later defaulting to False

daniel-sanche committed:
default begin_later to False

daniel-sanche committed:
Merge branch 'main' into new_transaction

daniel-sanche committed:
Merge branch 'main' into new_transaction

daniel-sanche committed:
Merge branch 'main' into new_transaction

renovate-bot committed:
chore(deps): update all dependencies (#505)

* chore(deps): update all dependencies

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

gcf-owl-bot[bot] committed:
build: update actions/checkout and actions/setup-python (#507)

Source-Link: https://github.com/googleapis/synthtool/commit/3551acd1261fd8f616cbfd054cda9bd6d6ac75f4
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:230f7fe8a0d2ed81a519cfc15c6bb11c5b46b9fb449b8b1219b3771bcb520ad2

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

danieljbruce commented on google/cloud/datastore/client.py:
```diff
@@ -122,7 +122,7 @@ def _extended_lookup(
     missing=None,
```
Just so I understand the architecture better. Is this function only called by read requests get, runQuery and runAggregationQuery?

daniel-sanche replied:
done

danieljbruce commented on google/cloud/datastore/client.py:
```diff
@@ -199,8 +199,22 @@ def _extended_lookup(
+    if transaction is not None:
+        transaction_id = transaction.id
+        if (
+            transaction_id is None
```
Is this check redundant? ie. if `transaction._status == transaction._INITIAL` is false then will `transaction_id is None` always be false? If so then we can remove this.

daniel-sanche replied:
I wouldn't expect the transaction to be in a state where the status is INITIAL and an id is set, but that's not enforced. So the question is what we want to do if we get in that state

- If we keep this check, it will fall back to its previous behaviour of just using the transaction id
- If we remove it, it will raise an exception, because it tries to set an id and create a new transaction in the same rpc

Thinking again, I think you might be right that it would be better to let it raise an exception in this situation. But it's not completely clear-cut, so I'll have another look at the context

danieljbruce commented on google/cloud/datastore/client.py:
```diff
@@ -199,8 +199,22 @@ def _extended_lookup(
+    )
     while loop_num < _MAX_LOOPS:  # loop against possible deferred.
         loop_num += 1
         request = {
```
How does this ensure that two requests are not being made at the same time? Or is the strategy here to error out with `Transaction already begun` on the second response instead of using a mutex to block one request while the other transaction request is in progress?

danieljbruce commented on google/cloud/datastore/client.py:
```diff
@@ -214,6 +228,10 @@ def _extended_lookup(
         )
 
+        # set new transaction id if we just started a transaction
+        if transaction and lookup_response.transaction:
```
We will want this business logic to work for runQuery and runAggregation query Gapic requests as well. It looks like this code is strictly only for Gapic Lookup requests? If so then this business logic should be in place for runQuery and runAggregation Query as well.

daniel-sanche replied:
done

danieljbruce commented on google/cloud/datastore/transaction.py:
```diff
@@ -304,6 +366,7 @@ def commit(self, retry=None, timeout=None):
             # Clear our own ID in case this gets accidentally reused.
             self._id = None
 
+    @_begin_if_not_began
```
Put and delete are not operations that make grpc calls. I am not sure we want to begin the transaction calls to these functions.

danieljbruce replied:
@bhshkh As discussed offline we will want to remove the call to begin transaction from the flowchart to match [the other diagram](https://docs.google.com/document/d/1iJqA82FD2jYidz6J4kloDlO7lQQ_jnJxDgTeGMa_0wA/edit?resourcekey=0-X7Sl0mXsOd7WQIrk8bHDRg&tab=t.0#bookmark=id.hetxait3zwlu) right?

daniel-sanche commented on google/cloud/datastore/helpers.py:
```diff
@@ -245,33 +247,35 @@ def get_read_options(eventual, transaction_id, read_time=None):
+    ]
+    if sum(is_set) > 1:
+        raise ValueError(
+            "At most one of eventual, transaction, or read_time is allowed."
```
Hmm interesting. My changes were based on the existing python implementation, so I haven't looked too much into the details here. But off the top of my head, it seems like:
- specifying an exact read_time snapshot seems to imply strong consistency. Otherwise, what does it mean to read from a specific timestamp?
- it doesn't make sense to associate an rpc with two different transactions, so it makes sense that transaction and new_transaction are incompatible
- using a transaction comes with it's own implicit read_time, so it makes sense that read_time, transaction, and new_transaction would all be mutually exclusive

I'll have to review the product docs around this. I'm curious how other languages handle this

danieljbruce replied:
Regarding your first bullet point, it looks like one of the benefits from eventual consistency is that the request will run faster as described in https://cloud.google.com/datastore/docs/concepts/structuring_for_strong_consistency#consistency_levels. So if the read time is a distant point in the past then the data is guaranteed to be available if strong or eventual consistency is used. In this case, allowing for eventual consistency with a specific read_time in the request may be advantageous for the user.

To answer your question, I think if the user runs a query with eventual consistency and a `read_time` then it reads the data from datastore's state at the read_time timestamp, but because it is eventually consistent it may run faster and not wait for other concurrent operations to complete. I see your point though as the docs say "Strongly consistent queries are transactionally consistent, meaning the results are based on a single, consistent snapshot of the data."

What do you think?

daniel-sanche replied:
Yeah, that's a good point. I'm curious if Kristen or Baha have any insights? Do other languages allow multiple options?

Otherwise I'll have to do some more research

daniel-sanche replied:
Actually I just looked at the ReadOptions proto, and this is [declared as a `oneof`](https://github.com/googleapis/googleapis/blob/0e50601ea3d1f828a90d2ddbd52920fcafd461fd/google/datastore/v1/datastore.proto#L563). So I think that gives us our answer

release-please[bot] committed:
chore(main): release 2.19.0 (#481)

Co-authored-by: release-please[bot] <55107282+release-please[bot]@users.noreply.github.com>

renovate-bot committed:
chore(deps): update dependency google-cloud-datastore to v2.19.0 (#508)

gcf-owl-bot[bot] committed:
build: update actions/upload-artifact and actions/download-artifact (#510)

Source-Link: https://github.com/googleapis/synthtool/commit/280ddaed417057dfe5b1395731de07b7d09f5058
Post-Processor: gcr.io/cloud-devrel-public-resources/owlbot-python:latest@sha256:346ab2efb51649c5dde7756cbbdc60dd394852ba83b9bbffc292a63549f33c17

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

daniel-sanche committed:
added new_transaction to query and aggregation classes

daniel-sanche committed:
got tests passing

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

daniel-sanche committed:
Merge branch 'main' into new_transaction

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

daniel-sanche committed:
remove begin from put and delete

daniel-sanche committed:
fixed jinja version

daniel-sanche committed:
fixed tests

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

daniel-sanche committed:
remove unused decorator

daniel-sanche committed:
removed unneeded delete override

danieljbruce commented on google/cloud/datastore/client.py:
```diff
@@ -199,8 +199,18 @@ def _extended_lookup(
+        transaction_id = transaction.id
+        if transaction._begin_later and transaction._status == transaction._INITIAL:
+            # if transaction hasn't been initialized, initialize it as part of this request
+            new_transaction_options = transaction._options
```
This code fragment is repeated multiple times. Do you think we could pull it into a function that returns new_transaction_options? Would that be a good idea to reduce the amount of repeated code?

Maybe something similar for transaction_id too?

daniel-sanche replied:
Sure, I added a new helper

danieljbruce commented on google/cloud/datastore/transaction.py:
```diff
@@ -274,6 +335,7 @@ def rollback(self, retry=None, timeout=None):
             # Clear our own ID in case this gets accidentally reused.
             self._id = None
 
+    @_abort_if_not_began
```
[This flowchart](https://docs.google.com/document/d/1iJqA82FD2jYidz6J4kloDlO7lQQ_jnJxDgTeGMa_0wA/edit?resourcekey=0-X7Sl0mXsOd7WQIrk8bHDRg&tab=t.0#bookmark=id.hetxait3zwlu) suggests that we should begin the transaction before making the commit call if it hasn't been started yet. Are you sure we shouldn't use begin_if_not_began?

daniel-sanche committed:
fixed commit without begin call

daniel-sanche committed:
improved tests

daniel-sanche committed:
refactored out helper function

daniel-sanche committed:
fixed lint

daniel-sanche committed:
added comment

daniel-sanche committed:
added tests for helper

daniel-sanche committed:
Merge branch 'main' into new_transaction

cindy-peng commented:
> This PR modifies the Transaction class to add a new `begin_later` argument.
> 
> When True, the transaction will not initialize itself at the start of the context manager `__enter__` block. Instead, it will only begin at the first rpc. If the first rpc is a read call, this allows us to create the transaction at the same time as the call, saving a network call.
> 
> This new behaviour is enabled by default, and is aligned with changes in other languages

Hi Daniel, it looks like `begin_later` is by default set to `False` so by default disabled? Do we want this behavior to be enabled by default?

daniel-sanche commented:
Yes, I believe it's intended to be off by default. Having it on or off may be more efficient depending on the circumstances, so we opted to keep it off by default (correct me if I'm misremembering @danieljbruce @bhshkh)

daniel-sanche committed:
Merge branch 'main' into new_transaction

daniel-sanche commented on google/cloud/datastore/batch.py:
```diff
@@ -218,7 +231,7 @@ def put(self, entity):
                  ``project`` does not match ours.
         """
-        if self._status != self._IN_PROGRESS:
+        if not self._allow_mutations():
```
The [current implementation](https://github.com/googleapis/python-datastore/blob/80612021b82d1b62d0271107f73571cde8262256/google/cloud/datastore/transaction.py#L374) behaves as you describe when the `begin_later` flag is True. When it's False, it currently falls back to the old expectations, where any action before calling begin is treated as an error (because begin was expected to be called as part of the init. If it wasn't, something is wrong)

It sounds like you're suggesting we should remove these extra validation checks when `begin_later=False`, and treat them both the same way after init time? If so, I don't really have any objections to that if that's more in line with other languages. Just want to make sure we're thinking about this the same way first

bhshkh replied:
Sorry for the confusion. I went through the diagram again and the current implementation is correct.
Resolving the conversation.

bhshkh commented on google/cloud/datastore/batch.py:
```diff
@@ -248,7 +261,7 @@ def delete(self, key):
                  ``project`` does not match ours.
         """
-        if self._status != self._IN_PROGRESS:
+        if not self._allow_mutations():
```
Sorry for the confusion. I went through the diagram again and the current implementation is correct.
Resolving the conversation.

daniel-sanche committed:
Merge branch 'main' into new_transaction
